File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string s) {
    int answer = s.length();  // 초기 답은 문자열의 길이로 초기화합니다.

    for (int i = 1; i <= s.length() / 2; ++i) {
        int pos = 0;            // 현재 압축을 시작할 위치를 나타냅니다.
        int length = s.length(); // 현재 압축된 문자열의 길이를 나타냅니다.

        // 현재 압축 단위로 문자열을 확인합니다.
        while (pos + i <= s.length()) {
            string unit = s.substr(pos, i);  // 압축 단위의 부분 문자열을 가져옵니다.
            pos += i;                        // 현재 단위 길이만큼 위치를 이동합니다.

            int cnt = 0; // 반복되는 부분 문자열의 개수를 세기 위한 변수입니다.

            // 현재 압축 단위로 반복되는 부분 문자열의 개수를 세줍니다.
            while (pos + i <= s.length()) {
                if (unit == s.substr(pos, i)) {
                    cnt += 1;
                    pos += i;
                } else {
                    break;
                }
            }

            // 만약 압축이 발생했다면 압축된 길이를 업데이트합니다.
            if (cnt > 0) {
                length -= i * cnt;

                // 압축된 횟수에 따라 길이를 조정해줍니다.
                if (cnt < 9) {
                    length += 1;
                } else if (cnt < 99) {
                    length += 2;
                } else if (cnt < 999) {
                    length += 3;
                } else {
                    length += 4;
                }
            }
        }

        // 현재까지의 최소 길이를 업데이트합니다.
        answer = min(answer, length);
    }

    return answer;
}

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    # 문자열 길이가 1인 경우 바로 1 반환
    if len(s) == 1:
        return 1
    
    # 가능한 모든 압축 결과 중 최소 길이를 저장
    min_length = len(s)
    
    # 1부터 문자열 길이의 절반까지 모든 가능한 단위로 시도
    for unit in range(1, len(s) // 2 + 1):
        compressed = ""
        count = 1
        prev = s[0:unit]
        
        # unit 크기만큼 문자열을 순회하며 압축
        for i in range(unit, len(s), unit):
            # 현재 단위만큼의 문자열 추출
            curr = s[i:i + unit]
            
            # 이전 문자열과 같다면 count 증가
            if prev == curr:
                count += 1
            # 다르다면 압축 결과에 추가하고 초기화
            else:
                if count > 1:
                    compressed += str(count)
                compressed += prev
                prev = curr
                count = 1
        
        # 마지막 문자열 처리
        if count > 1:
            compressed += str(count)
        compressed += prev
        
        # 최소 길이 갱신
        min_length = min(min_length, len(compressed))
    
    return min_length